### ISSUE TITLE
Chained table / array declarations

### ISSUE THREAD

**Issue by meghprkh** [toml-lang/toml#1019]
This might be a slightly niche request but currently TOML requires writing out the full declaration per line.

My proposal is to keep the same explicitness, but allow chaining them, such that

```toml
[[request.test]][capture.jsonpath]
id = '$.id'
```

is equivalent to 

```toml
[[request.test]]
[request.test.capture.jsonpath]
id = '$.id'
```

Or in JSON 

```json
{
  "request": {
    "test": [
      {
        "capture": {
          "jsonpath": {
            "id": "$.id"
          }
        }
      }
    ]
  }
}
```

In general, the problem I feel this addresses is alteration of array and dictionary objects when declaring structures. This might not be super common but deeply nested JSONs are convenient sometimes.

**Comment by eksortso** [comment]
I'll have two responses. This one discusses ways to do the things that the proposed syntax is intended to solve, using existing or upcoming syntax. I'll make another response later to address the proposal directly and bring up other syntaxes brought up in the past. I have some background experience here, so I have some opinions to share. But I'll write about those things later.

There is already a way to define subtables within a single line to avoid mixing up array and table syntax, and that is with the use of dotted keys. The example that you provide can be written this way:

```toml
[[request.test]]
capture.jsonpath.id = '$.id'
```

When TOML v1.1.0 is released, it will allow inline tables to span multiple lines, allowing for deep nesting to be expressed in a way that is familiar to users accustomed to nested formats like JSON. The following will be valid after the release of TOML v1.1.0rc1:

```toml
[[request.test]]
capture = {
    jsonpath = {
        id = '$.id'
    }
}
```

**Comment by ChristianSi** [comment]
Yeah I suppose dotted keys are the more readable way to write this, and they don't require any new syntax.

**Comment by levicki** [comment]
> The following will be valid after the release of TOML v1.1.0rc1:

That's horrible and you should be ashamed of yourself.

**Comment by levicki** [comment]
> @levicki Don't moralize.

I'll stop morallzing when people with the Borg mindset like you stop ruining everything they touch with their "improvements".

They not only lack the understanding of why something was made in a certain way to begin with &mdash; they don't respect the original design enough to care and instead just want to trample through it and change it to suit whatever workflow they had in mind no matter how disconnected from the original usage intent that might be.

An example of such people would be everyone clamoring for Apple to allow sideloading and alternate stores on iOS. The argument they make (that I won't have to enable sideloading and that I will still be able to buy stuff in the single store) is as disingenious as yours about me being able to still use the old TOML format.

> As it stands, there are enough people nowadays whose whole life experience involved XML and JSON, and never any flavor of INI.

And they can all happily continue to use XML and JSON while I can happily continue to use TOML.

Who is forcing those people to use TOML if they want it to be JSON formatted? I mean, do they also go to a mosque and demand to hear gospel or go into a kosher store and ask to buy pork? It's ridiculous.

> Fact is, TOML is still tightly connected to the INI approach of doing things. That will never change.

Once you start nesting things in those hideous braces it won't be connected to its original design goal.

You all seem to have forgotten this part of the spec:

> TOML **prioritizes humans**
> TOML aims to be a **minimal** configuration file format that:
> is **easy to read due to obvious semantics**
> **maps unambiguously to a hash table**
> is **easy to parse** into data structures in a wide variety of languages

Go ahead and explain how making it more JSON-like moves it closer to that goal.

**Comment by eksortso** [comment]
Before you insult me any further, @levicki, do note that I complained about TOML turning into JSON for years. Literally years. [This comment](https://github.com/toml-lang/toml/issues/516#issuecomment-884641627) should make my case for me. Anyway, that entire thread is worth reading, even if you hated its outcome.

That's ancient history now. And you're upset right now that I suggested the JSON-like syntax as a potential _secondary_ solution. I didn't say it was ideal, just that it would be possible.

I'm sorry that TOML is not the idealized solution, encased in ice forever, that you think it should be. Formats for humans sometimes bend to human needs, whether we like it or not. Complain to your colleagues who use features you don't like, not us, and certainly not me.

But also look at any `cargo.toml` or `pyproject.toml` file in production code, and realize that these common use cases adhere closely to the INI spirit. Also note that these files are typically machine-generated by other building tools.

**Comment by eksortso** [comment]
Before you insult me any further, @levicki, do note that I complained about TOML turning into JSON for years. Literally years. [This comment](https://github.com/toml-lang/toml/issues/516#issuecomment-884641627) should make my case for me. Anyway, that entire thread is worth reading, even if you hated its outcome.

That's ancient history now. And you're upset right now that I suggested the JSON-like syntax as a potential _secondary_ solution. I didn't say it was ideal, just that it would be possible.

I'm sorry that TOML is not the idealized solution, encased in ice forever, that you think it should be. Formats for humans sometimes bend to human needs, whether we like it or not. Complain to your colleagues who use features you don't like, not us, and certainly not me.

But also look at any `cargo.toml` or `pyproject.toml` file in production code, and realize that these common use cases adhere closely to the INI spirit. Also note that these files are typically machine-generated by other building tools.

**Comment by levicki** [comment]
I have no intention of insulting anyone, but if you at some point have defended TOML from these changes you seem to have joined the horde now.

This particular sentence in the very first response to your bitter hyperbolic comment stands out however, and it was left unaddressed:

> I don't think anyone has forgotten, **they just disagree with the intentions behind the design. The design is not sacrosanct** and it should not be treated as such.

It actually proves the point I made above word for word &mdash; (I hate to quote myself but this time I have to):

> They not only seem to **lack the understanding of why something was made in a certain way to begin with — they don't seem to respect the original design enough to care** and instead just want to trample through it and change it to suit whatever workflow they had in mind no matter how disconnected from the original usage intent that might be.

If I am wrong and you actually still disagree with their way of thinking that everything needs fixing and their toxic attitude that nothing is holy then don't give them any legitimacy by participating in what they are doing, much less try to defend it.

**Comment by meghprkh** [comment]
The dotted syntax would not allow arbitrary nesting like (not a great example but hopefully conveys the motivation):

```ini
[[planets]]
name = "Earth"

[[planets]][fauna][[categories]]
name = "mammal"
description = "xyz"

[[planets]][fauna][[categories]]
name = "reptile"
description = "xyz"

[[planets]]
name = "Mars"
fauna = {}
```

```yaml
planets:
  - name: Earth
    fauna:
      categories:
        - name: mammals
          description: xyz
        - name: reptiles
          description: xyz
  - name: Mars
    fauna: {}
```

It should be worked out on whether this is going to be a config language that doesn't support well nested data and a small subset of it or a generic data language.

**Comment by arp242** [comment]
I do not find it easy to follow that last example; I think this is much clearer:

	[[planets]]
	name  = "Earth"
	fauna = {
		categories = [
			{name = "mammals",  description = "xyz"},
			{name = "reptiles", description = "xyz"},
		],
	}

	[[planets]]
	name  = "Mars"
	fauna = {}

**Comment by levicki** [comment]
> The dotted syntax would not allow arbitrary nesting...

The real question is whether **every** config file format must support that?

In my opinion, they shouldn't &mdash; JSON and XML are much better suited for nesting than any other format and they should be used instead of TOML if you need nesting to that extent.

TOML wasn't designed to support it and shoehorning it in after the fact is just going to make it worse for its current use.

To me, TOML is more akin to a database instead of a tree structure and people have successfully stored all sorts of data in databases which don't support nesting for decades.

> I do not find it easy to follow that last example; I think this is much clearer...

I think that both of you don't understand existing TOML specification. If you want a contrived example of deep nesting, here's one that shows how dotted keys already allow flattened representation of nesting:
```
[[Planets]]
Name = "Earth"

[[Planets.Inhabitants]]
Name = "Humans"
Description = "xyz"
Pets = [
	{ Name = "Dogs", Description = "xyzzy" },
	{ Name = "Cats", Description = "xyzzy" },
]

[[Planets]]
Name = "Mars"

[[Planets.Inhabitants]]
Name = "Little Gray Men"
Description = "xyz"
Pets = [
	{ Name = "Martian Brain Slug", Description = "xyzzy" },
]
```
To me, that's way more readable (and easier to edit!) than the same thing expressed as JSON:
```
{
        "Planets": [
                {
                        "Name": "Earth",
                        "Inhabitants": [
                                {
                                        "Name": "Humans",
                                        "Description": "xyz",
                                        "Pets": [
                                                {
                                                        "Name": "Dogs",
                                                        "Description": "xyzzy"
                                                },
                                                {
                                                        "Name": "Cats",
                                                        "Description": "xyzzy"
                                                }
                                        ]
                                }
                        ]
                },
                {
                        "Name": "Mars",
                        "Inhabitants": [
                                {
                                        "Name": "Little Gray Men",
                                        "Description": "xyz",
                                        "Pets": [
                                                {
                                                        "Name": "Martian Brain Slug",
                                                        "Description": "xyzzy"
                                                }
                                        ]
                                }
                        ]
                }
        ]
}
```

**Comment by ChristianSi** [comment]
@levicki I very much agree that the TOML example you give is more readable than its JSON equivalent – and I'm also very sure that nobody would be use TOML *just* to mimic JSON. After all, what would be the point of that?

I'm nevertheless convinced that the change we made in TOML 1.1 is for the better, since your example is of course a toy example. Made a bit more realistic, and shortened to just the relevant bits, it might look rather like this:

```toml
[[Planets.Inhabitants]]
Pets = [
    { Name = "Dogs", Description = "Mammals of the species Canis familiaris, of highly variable appearance because of human breeding." },
    { Name = "Cats", Description = "Mammals of the domesticated species Felis catus, commonly kept as a house pets." },
]

[[Planets.Inhabitants]]
Pets = [
    { Name = "Martian Brain Slug", Description = """A Martian gastropod mollusc, popular as pet despite being a bit challenging to keep,
as it likes to feed on its owners' brains. Many Martian houses have a warning sign
"Beware of the slug" to warn clueless passers-by to take the necessary precautions.""" },
]
```

Now, since these descriptions are quite long (one of them even spans multiple lines), a TOML editor might naturally feel inclined to insert a linebreak before them:

```toml
[[Planets.Inhabitants]]
Pets = [
    { Name = "Dogs",
      Description = "Mammals of the species Canis familiaris, of highly variable appearance because of human breeding."
    },
    { Name = "Cats",
      Description = "Mammals of the domesticated species Felis catus, commonly kept as a house pets."
    },
]

[[Planets.Inhabitants]]
Pets = [
    { Name = "Martian Brain Slug",
      Description = """A Martian gastropod mollusc, popular as pet despite being a bit challenging to keep,
as it likes to feed on its owners' brains. Many Martian houses have a warning sign
"Beware of the slug" to warn clueless passers-by to take the necessary precautions."""
    },
]
```

However, TOML 1.0 doesn't allow this, which is a very surprising and non-intuitive restriction. Even more so, since TOML has always allowed line breaks in arrays, which are otherwise very closely related to inline tables.

Fortunately, TOML 1.1 removes this well-intended but surprising restriction, allowing people to add linebreaks more freely without causing accidental breakage. Hurrah!

**Comment by levicki** [comment]
@ChristianSi 
> a TOML editor might naturally feel inclined to insert a linebreak before them

Maybe an intellectually challenged TOML editor, not someone with common sense.

> However, TOML 1.0 doesn't allow this, **which is a very surprising and non-intuitive restriction**.

Of course it doesn't allow it for Christ sake &mdash; it isn't called an **inline** table for nothing. You know what [inline](https://www.merriam-webster.com/dictionary/in-line) means in English, don't you? What's so surprising and non-intuitive about an inline table not being allowed to span multiple lines?

> Fortunately, TOML 1.1 removes this well-intended but surprising restriction, allowing people to add linebreaks more freely without causing accidental breakage. Hurrah!

Yes, and now anyone skimming this description:
```
[[Planets.Inhabitants]]
Pets = [
    { Name = "Martian Brain Slug",
      Description = """A Martian gastropod mollusc, popular as pet despite being a bit challenging to keep,
as it likes to feed on its owners' brains. Many Martian houses have a warning sign
"Beware of the slug" to warn clueless passers-by to take the necessary precautions."""
    },
]
```
Will have to scan for the start and end of that whole description string and match the `"""` in order to figure out whether that comma on the end of first line is part of the string or part of the table. Bonus points? It looks as ugly as JSON now that the braces are not on the same line and that you can have both brackets and braces on different lines which before you couldn't.

I am sorry, but your idea, even if you believe it to be well intentioned, is bad and it comes from your poor understanding of things.

TL;DR is that TOML doesn't need "fixing" because it isn't broken but very sensibly designed &mdash; it just needs users who understand it's design goals and who will recognize when they need to drop it and use a different config format instead of trying to shoehorn things like this into it. Nowhere does it say you have to use TOML for every config file -- just leave the ugly stuff like that for `![CDATA[]]` sections in XML and allow TOML to remain elegant instead of making it as ugly as the others.

**Comment by arp242** [comment]
> Maybe an intellectually challenged TOML editor, not someone with common sense who understands the specification.

Stop it with these insults; this is really crossing the line.

**Comment by levicki** [comment]
> Stop it with these insults; this is really crossing the line.

And how do you propose calling people who seemingly don't understand that "inline" means "arranged on a single line", who decide to change a well thought out specification based on said lack of understanding, and who appear happy and boastful about their own ignorance?

I am always eager to improve my social skills and vocabulary so please advise.

**Comment by arp242** [comment]
You don't have to call people anything; just state your case why you think A is better than B. It's not hard.

**Comment by levicki** [comment]
> You don't have to call people anything; just state your case why you think A is better than B. It's not hard.

Very well, I have edited the post to exclude what you quoted as offending.

I hope you will now be able to focus on the arguments I brought up against making this change.

**Comment by arp242** [comment]
There isn't much to address because all you've said is "I don't like it" in different ways. I can't argue with that.

All I can say I do like it. And that many people like it. And that #516 is the most upvoted issue on this repo. And "I'd love to use TOML if it would allow newlines" comes up in most Hacker News discussions and such that I've seen. And a number of people have asked "when is TOML 1.1 coming out because I'd like this feature" over the last year or so (which they haven't done for any other change, AFAIK).

So your view is a distinct minority, and the only thing that could possibly change the course is some practical real-world issue. That is: ambiguous syntax, it's hard to implement, hard to give good errors. Things like that.

**Comment by levicki** [comment]
> There isn't much to address because all you've said is "I don't like it" in different ways. I can't argue with that.

Now you are being disingenious (i.e. you are lying that there's nothing to address).

You only need to address this:

TOML specification declares an inline table syntax which, unlike the other available table syntax, is supposed to be written on a single line. You want to change what inline table means in TOML specification, which also requires disregarding what "inline" means in English. How does that even make sense to you? Please explain, can't wait to hear your rationalization.

> All I can say I do like it. And that many people like it. And that #516 is the most upvoted issue on this repo.

Nobody cares what you like if it doesn't make sense and if it requires distorting reality and throwing common sense out of the window to accomplish it.

> And "I'd love to use TOML if it would allow newlines" comes up in most Hacker News discussions and such that I've seen.

See, that's the behavior I was talking about.

People who say "I'd love to use iPhone if it would allow sideloading (and 3rd party stores, and if I didn't have to pay Apple tax, and if the battery was removable, and if it had external storage, and 3.5 mm headphone jack, and...)".

The truth is &mdash; if they really wanted to use it they would have found a way to make it work for them by now. As it stands, they are only attention seekers who are trying to appear relevant by making unreasonable demands.

Nobody should give a rat's arse whether you want to use TOML or not, especially if your condition for using it is to use it improperly and worse yet if you want this improper way to be the official way &mdash; it's a file format, not a popularity contest.

> So your view is a distinct minority

Have you actually tried running a poll with current TOML users to see how many would approve of this change?

> and the only thing that could possibly change the course is some practical real-world issue.

The real-world issues are:

1. I have shown how existing syntax works just fine for all the contrived examples given here. You still need to demonstrate what problem this change solves other than "those people won't use TOML unless we implement it".
2. You still need to figure out a new name for inline tables so your proposal doesn't clash with English language.

I understand it is hard to look back at something you put considerable effort into, admit you were wrong, and discard it. But unless you can give some sensible non-dismissive answers to my points above that's exactly what should happen.

**Comment by eksortso** [comment]
Can we get back on topic? I think we may be able to accomplish the same thing that @meghprkh wants with a much simpler syntax addition to TOML. This is a refinement of an [earlier idea that I floated years ago](https://github.com/toml-lang/toml/issues/309#issuecomment-361671675), which I think is worth revisiting.

In short, why don't we allow an empty pair of square brackets `[]` after names of table arrays, but immediately after their part of the dotted key? They can effectively be ignored, unless they're used after a name that's not declared to be an array of tables, or if they appear at the very end of the table name.

Going back to your original idea, @meghprkh, consider your example:

```toml
# Valid in TOML v1.0.0
[[request.test]]
[request.test.capture.jsonpath]
id = '$.id'
```

Since `test` is a table array, why not allow this, so that readers can track the arrays more easily. They can be ignored when used in the middle of a long name. This is akin to the original proposal's "`[[request.test]][capture.jsonpath]`".

```toml
# Part 1 of proposal: Make `[request.test[].capture.jsonpath]` equivalent to `[request.test.capture.jsonpath].
[[request.test]]
[request.test[].capture.jsonpath]  # NOTE: `test` is already known to be an array of tables.
id = '$.id'
```

And then, to be consistent, what if we made `[request.test[]]` equivalent to `[[request.test]]`?

```toml
# Part 2 of proposal: Make `[request.test[]] equivalent to `[[request.test]]`.
[request.test[]]  # NOTE: this marks a new element of a table array.
[request.test[].capture.jsonpath]
id = '$.id'
```

This proposal is fully compatible with the current syntax and would only be valid inside table headers and array-of-table headers. Both old and new types of syntax may be used, and the equivalences mentioned above can be honored if mixed.

@meghprkh What do you think of this way of cleaning up the confusion that can surround mixed table and array declarations? How would you say it compares to your proposal?

**Comment by arp242** [comment]
I find this exceedingly complex and non-obvious. Few people will be able to look at:

    [[request.test]]
    [request.test[].capture.jsonpath]
    id = '$.id'

And intuitively be able to understand what this means.

This is about is not having to write the full table name in:

    [[request.test]]
    [request.test.capture.jsonpath]
    id = '$.id'

And your example doesn't avoid that, and is actually *longer*.

**Comment by levicki** [comment]
> This is about is not having to write the full table name in:
> 
> ```
> [[request.test]]
> [request.test.capture.jsonpath]
> id = '$.id'
> ```
> 
> And your example doesn't avoid that, and is actually _longer_.

While I agree that his example is even worse, you don't really have to write the full table name, `jsonpath` can be written as an inline table (because that's what they are for):
```
[request]
[[request.test]]
[request.test.capture]
jsonpath = { id = "$.id" }
```
Note that even with having to write necessary tables above the final one (which you will certainly have to write at least once if your data has arrays) it is still shorter, more readable, and easier to write than the equivalent JSON:
```
{
    "request": {
        "test": [
            {
                "capture": {
                    "jsonpath": {
                        "id": "$.id"
                    }
                }
            }
        ]
    }
}
```
In short, dotted keys and inline tables are used to balance the data tree and keep it as flat as possible.

In my opinion, the concept works very well as-is, and changing the inline table as suggested would break it.

Finally, having to write nested `{}` like in JSON is way worse than just selecting and copy/pasting a whole line.

**Comment by eksortso** [comment]
@arp242 I never recommended doing this without adopting both sections. You didn't comment on the Part 2 example.

Just because I show something that's possible doesn't mean I endorse it. Why does this keep happening??

**Comment by eksortso** [comment]
We're not playing configuration golf; we don't necessarily need the shortest example. The dotted-key approach gave us a short example. But the clarity of that approach is up for debate, especially if there's more to the subtables `capture` and `jsonpath` that we don't know about.

**Comment by eksortso** [comment]
One thing that I admire in @meghprkh's proposal is that it is capable of indicating a new table element on the same line as its subtable, using the `[[aot]]` notation at the beginning. That's clear. 

But what I do _not_ like about it is that a _relative_ `[subtable]` path inside brackets suddenly appears. Even though the brackets butt up against each other on the same line, I still consider this approach unintuitive. Table headers must, and always should, be absolute paths.

And two bracketed names pushed together don't immediately suggest an absolute path, at least in my eyes. This is partly why I reintroduced the empty-bracket proposal; that approach makes table array elements stand out in the table's name, which will reduce confusion surrounding nested arrays and tables found together.

So, if there's any other way to combine a new array element with its subtable (or sub-array) on a single line without ambiguity, I'd appreciate it.

**Comment by arp242** [comment]
> @arp242 I never recommended doing this without adopting both sections. You didn't comment on the Part 2 example.

To be honest I find the second part even worse.

We already have [..] table declarations, dotted keys, and inline arrays and tables. I think that's enough.

**Comment by levicki** [comment]
> We're not playing configuration golf

Really? What's all those poorly written contrived examples (which I rewrote to fully utilize TOML abilities) advocating for the change then?

> we don't necessarily need the shortest example.

The whole point of TOML is to write the shortest possible human readable configuration file. I am sorry you can't see that.

> The dotted-key approach gave us a short example. But the clarity of that approach is up for debate, especially if there's more to the subtables `capture` and `jsonpath` that we don't know about.

There might be more to those sub-tables, but I don't think we are supposed to imagine what wasn't shown explicitly and much less cater to it.

If someone needs deeper nesting mixed with multi-line data which makes using current TOML specification impractical and / or cumbersome, then they have simply chosen the wrong configuration file format for the task at hand.

**Comment by eksortso** [comment]
And I am sorry that I won't cave to your ad hominem attacks. Where are you getting "shortest possible" from?? That was never a goal. That's just a consequence of our overall goal.

Human readability is our overall goal. Obviousness  and minimalism are our driving purposes. But since it's possible for the minimal to lose coherence, we've added features to make the language more convenient. These were never just thrown in at random.

Like it or not, "short" has always been an afterthought. But it came from deliberate choices made over the years. Because human beings happen to like "short" configurations.

**Comment by arp242** [comment]
> > We already have [..] table declarations, dotted keys, and inline arrays and tables. I think that's enough.
>
> You say this, but there is nothing in my empty-bracket proposal that goes beyond the concepts of [table] and [[aot]] declarations.

But it's a new way to define those, no? I just don't see what this adds over the existing methods to define them. One could perhaps find it a bit easier or better, and that's fair, but the existing methods aren't really that cumbersome, so I don't see what yet another way adds.

**Comment by levicki** [comment]
> And I am sorry that I won't cave to your ad hominem attacks.

What about me saying that I am genuinely sorry about you not seeing something that is obvious to me looks like an ad hominem attack? No wait, don't tell me, because that wasn't an ad hominem attack at all. You are just looking for excuses to discard my arguments.

> Where are you getting "shortest possible" from??

It's in the format name:

![image](https://github.com/toml-lang/toml/assets/16415478/f50920ba-e33c-489f-a636-9aad0777a03e)

And in the blurb:

_TOML aims to be **a minimal configuration file format** that's easy to read due to obvious semantics. TOML is designed to map unambiguously to a hash table. TOML should be easy to parse into data structures in a wide variety of languages._

To me "minimal" and "shortest possible" are synonyms in this context.

> That was never a goal.

If it wasn't then it wouldn't be in the name. It's just that Tom's goal seems to have been different than yours.

> Human readability is our overall goal.

Adding new lines into an inline table which uses curly braces to allow nesting (and indentation) does **not** improve readability.

In TOML, nesting and indentation is already used for arrays:
```
[Blob]
Data = [
    0x44, 0x4f, 0x53, 0x00
]
```
And that's what makes those arrays obvious when you look at the file. Add another element that can be nested (curly braces) and you will break that obviousness.

> Obviousness and minimalism are our driving purposes.

You may claim so, but based on the change proposed and the stubborness associated with pushing it through despite valid counter-arguments presented I am not seeing it.

I wonder... would something like this be acceptable instead?
```
[request]
[[request.test]]
[>capture.jsonpath]
id = '$.id'
[>torture.jsonpath]
junk = '$.junk'
```
Here, `>` is a shorthand for last defined array start path (`request.test`) so you don't have to write it out fully.

Granted, that would prevent people from using `>` (or whatever other symbol is picked) inside quoted dotted key but you can't have everything.

**Comment by eksortso** [comment]
I do like the idea of `[...` as a mechanism for abbreviating super-table names. Ellipses for an abbreviation of the prefix makes sense. It does count as an addition to the language, but it's an elegant way of doing it. It requires no new special characters. Please mention it again in #1017, because it would serve well as a potential new "alias" syntax!

**Comment by ChristianSi** [comment]
I guess the obvious question re the last proposal is: Why just refer back to the last array of table element, without also offering a way to refer to earlier (non-array) table names? My proposal, rediscovered by @eksortso, was meant to be equally usable for both.

**Comment by eksortso** [comment]
@ChristianSi As it turns out, this fits into my argument against @levicki's take that our use of the word "minimal" somehow refers not to the brevity of TOML's syntax, but rather to the length of the documents that TOML users generate, which I claim is essentially wrong.

In the past, although we've discussed all distinct types of syntax for TOML, only the ones that succinctly fit a specific need are adopted. Past discussions make that drive towards minimalism perfectly clear to understand; although we may like a new syntax, objections are raised when they're too difficult to explain concisely, and more complicated ones are eventually abandoned.

Read #1017, #744, and back even further for examples of how to make long table names shorter. The intention isn't to make the documents shorter, so much as it is to make them more convenient for their writers. This illustrates where the drive towards minimalism comes into play.

@ChristianSi, I no longer think that we should attempt to adopt parent-directory-style notation. The less that writers have to be aware of table and array depth while writing, the better. The alias proposals so far under #1017 take one name (the previous explicitly defined absolute table name) and only expand from there. That idea is extremely simple to grasp, and it's my favorite going forward. That's minimalism: we prefer simpler ideas that reach the same goals.

The way to express these "aliases" involved special characters to indicate when to use an alias. We started with explicitly written alias names. Then we switched to a "prefix" approach; one version using `&` was very elegant. Then `...` ellipses were suggested, and that appealed to me more because:

* it didn't require a new character,
* its use as abbreviation has precedence in other uses,
* and it stands out a lot more than a single `&` or `@` would.

It's elegant. It's not shorter since it uses three characters instead of just one. But it is the most elegant approach and appears to be more obvious.

That's how we apply a _minimal_ aesthetic. Take simpler ideas and more understandable syntax, then shape it so that users may learn by example, so they can use the new syntax. Make fewer hurdles to jump over in order to understand.

**Comment by levicki** [comment]
> I guess the obvious question re the last proposal is: Why just refer back to the last array of table element, without also offering a way to refer to earlier (non-array) table names? My proposal, rediscovered by @eksortso, was meant to be equally usable for both.

TOM specification says that:

 - You must have a created array before adding a table to it
 - Entries following an array refer to that array

What I proposed fits into those rules and allows shortening dotted keys which is what I figure was the core complaint of most people wanting to do deep nesting in TOML and who were suggesting various ways like using aliases, path semantics, or breaking inline tables using new lines.

As a bonus, ellipsis already has a universal meaning outside of TOML and is thus more intuitive for people who aren't tech-savvy.

On the other hand, having a variable number of dots would make it non-intuitive and error-prone because instead of one or three dots you can have any number of them and you have to count them.

Refering to other arbitratry elements in the path would require much more complex specification (and thus parser) changes so that's why I stuck with what is basically already in the specification (sans the shortening itself being allowed).

> As it turns out, this fits into my argument against @levicki's take that our use of the word "minimal" somehow refers not to the brevity of TOML's syntax, but rather to the length of the documents that TOML users generate, which I claim is essentially wrong.

> The intention isn't to make the documents shorter, so much as it is to make them more convenient for their writers.

I do agree with most of your other points but this is arguing semantics &mdash; the length of TOML documents is directly proportional to the brevity of its syntax, and making them shorter is making them more convenient for writers and readers. You can't have one without the other and that's what I am arguing here.

**Comment by ChristianSi** [comment]
@eksortso , @levicki: Yeah, sure, I think I could come on board regarding the ellipsis proposal too.

**Comment by pradyunsg** [comment]
Apologies for the slow response from my end here.

@levicki has a 7 day block from interacting on this organisation for their behaviour here ~3 weeks ago. In the future, please be respectful of differing viewpoints and avoid using derogatory language when referencing to others or their work.

Jumping back to OP again:

> This might be a slightly niche request but currently TOML requires writing out the full declaration per line.

This is an intentional design choice.

I don't think TOML should allow chaining tables/array of tables on a single line with `][` in the line.

---

Further discussion on ellipsis-related proposals/other shorthands can take place at #1017. Closing this out to reflect that the model that OP proposes isn't going to something TOML is going to evolve in the direction of today.
